python-package/lightgbm/basic.py

```diff
@@ -3079,7 +3079,14 @@ def __boost(self, grad, hess):
         assert grad.flags.c_contiguous
         assert hess.flags.c_contiguous
         if len(grad) != len(hess):
-            raise ValueError(f"Lengths of gradient({len(grad)}) and hessian({len(hess)}) don't match")
+            raise ValueError(f"Lengths of gradient ({len(grad)}) and hessian ({len(hess)}) don't match")
+        num_train_data = self.train_set.num_data()
+        num_models = self.num_model_per_iteration()
```

Comment:
Use precomputed at the init phase value for the speedup.
https://github.com/microsoft/LightGBM/blob/67b4205c8043326553e294fa2c01ad1189784631/python-package/lightgbm/basic.py#L2645-L2649
```suggestion
        num_models = self.__num_class
```